File: src/Crawler/Options/Group.php
<?php

/*
 * This file is part of the SiteOne Crawler.
 *
 * (c) Ján Regeš <jan.reges@siteone.cz>
 */

declare(strict_types=1);

namespace Crawler\Options;

use InvalidArgumentException;

class Group
{

    /**
     * Unique application code for the group
     * @var string
     */
    public readonly string $aplCode;

    /**
     * Readable name for the group
     * @var string
     */
    public readonly string $name;

    /**
     * Array of options - key is property name to fill
     * @var Option[]
     */
    public readonly array $options;

    /**
     * @param string $aplCode
     * @param string $name
     * @param Option[] $options
     * @throws InvalidArgumentException
     */
    public function __construct(string $aplCode, string $name, array $options)
    {
        foreach ($options as $option) {
            if (!($option instanceof Option)) {
                throw new InvalidArgumentException('Options must be instance of Option class');
            }
        }
        $this->aplCode = $aplCode;
        $this->name = $name;

        $optionsWithKeys = [];
        foreach ($options as $option) {
            $optionsWithKeys[$option->propertyToFill] = $option;
        }
        $this->options = $optionsWithKeys;
    }

}

File: src/options/group.rs
// SiteOne Crawler - Option group for organizing options
// (c) Jan Reges <jan.reges@siteone.cz>
//

use indexmap::IndexMap;

use super::option::CrawlerOption;

#[derive(Debug, Clone)]
pub struct OptionGroup {
    /// Unique application code for the group
    pub apl_code: String,

    /// Readable name for the group
    pub name: String,

    /// Options indexed by property_to_fill name
    pub options: IndexMap<String, CrawlerOption>,
}

impl OptionGroup {
    pub fn new(apl_code: &str, name: &str, options: Vec<CrawlerOption>) -> Self {
        let mut options_map = IndexMap::new();
        for option in options {
            options_map.insert(option.property_to_fill.clone(), option);
        }

        Self {
            apl_code: apl_code.to_string(),
            name: name.to_string(),
            options: options_map,
        }
    }
}
